File: Bioinfo/src/main/java/codes/java/SW.java
package codes.java;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.util.Locale;
public class SW {
    public static void smithWaterman(String seq1, String seq2) {
        long startTime = System.currentTimeMillis();
        int matchScore = 1;
        int mismatchPenalty = -1;
        int gapPenalty = -1;
        int[][] scoreMatrix = new int[seq1.length() + 1][seq2.length() + 1];
        int maxScore = 0;
        int[] maxPos = new int[2];
        for (int i = 1; i <= seq1.length(); i++) {
            for (int j = 1; j <= seq2.length(); j++) {
                int match = scoreMatrix[i - 1][j - 1] + score(seq1.charAt(i - 1), seq2.charAt(j - 1), matchScore, mismatchPenalty);
                int delete = scoreMatrix[i - 1][j] + gapPenalty;
                int insert = scoreMatrix[i][j - 1] + gapPenalty;
                scoreMatrix[i][j] = Math.max(0, Math.max(match, Math.max(delete, insert)));
                if (scoreMatrix[i][j] > maxScore) {
                    maxScore = scoreMatrix[i][j];
                    maxPos[0] = i;
                    maxPos[1] = j;
                }
            }
        }
        int i = maxPos[0];
        int j = maxPos[1];
        int gaps = 0;
        while (scoreMatrix[i][j] != 0) {
            if (scoreMatrix[i][j] == scoreMatrix[i - 1][j - 1] + score(seq1.charAt(i - 1), seq2.charAt(j - 1), matchScore, mismatchPenalty)) {
                i--;
                j--;
            } else if (scoreMatrix[i][j] == scoreMatrix[i - 1][j] + gapPenalty) {
                gaps++;
                i--;
            } else {
                gaps++;
                j--;
            }
        }
        long endTime = System.currentTimeMillis();
        DecimalFormatSymbols symbols = new DecimalFormatSymbols(Locale.US);
        DecimalFormat df = new DecimalFormat("0.00", symbols);
        double executionTime = (endTime - startTime) / 1000.0;
        System.out.println(df.format(executionTime));
        System.out.println(maxScore);
        System.out.println(gaps);
        System.out.println("41"); // Ajuste conforme necessário
    }
    public static int score(char char1, char char2, int matchScore, int mismatchPenalty) {
        return (char1 == char2) ? matchScore : mismatchPenalty;
    }
    public static void main(String[] args) {
        String seq1 = args[0];
        String seq2 = args[1];
        smithWaterman(seq1, seq2);
    }
}

File: Bioinfo/src/main/java/codes/python/SW.py
import time
import sys
import math
def score(char1, char2):
    if char1 == char2:
        return 1
    else:
        return -1
def calculate_evalue(score, m, n, lambda_value):
    K = 0.1
    return K * m * n * math.exp(-lambda_value * score)
def smith_waterman(seq1, seq2, match_score=1, mismatch_penalty=-1, gap_penalty=-1, lambda_value=0.0):
    inicio = time.perf_counter()
    score_matrix = [[0] * (len(seq2) + 1) for _ in range(len(seq1) + 1)]
    max_score = 0
    max_pos = (0, 0)
    for i in range(1, len(seq1) + 1):
        for j in range(1, len(seq2) + 1):
            match = score_matrix[i - 1][j - 1] + score(seq1[i - 1], seq2[j - 1])
            delete = score_matrix[i - 1][j] + gap_penalty
            insert = score_matrix[i][j - 1] + gap_penalty
            score_matrix[i][j] = max(0, match, delete, insert)
            if score_matrix[i][j] > max_score:
                max_score = score_matrix[i][j]
                max_pos = (i, j)
    aligned_seq1 = aligned_seq2 = barrinha = ''
    i, j = max_pos
    while score_matrix[i][j] != 0:
        if score_matrix[i][j] == score_matrix[i - 1][j - 1] + score(seq1[i - 1], seq2[j - 1]):
            aligned_seq1 = seq1[i - 1] + aligned_seq1
            aligned_seq2 = seq2[j - 1] + aligned_seq2
            if score(seq1[i - 1], seq2[j - 1]) == 1:
                barrinha = '|' + barrinha
            else:
                barrinha = ':' + barrinha
            i -= 1
            j -= 1     
        elif score_matrix[i][j] == score_matrix[i - 1][j] + gap_penalty:
            aligned_seq1 = seq1[i - 1] + aligned_seq1
            aligned_seq2 = "-" + aligned_seq2
            barrinha = '-' + barrinha
            i -= 1
        else:
            aligned_seq1 = "-" + aligned_seq1
            aligned_seq2 = seq2[j - 1] + aligned_seq2
            barrinha = '-' + barrinha
            j -= 1
    fim = time.perf_counter()
    tempo_execucao = fim - inicio
    evalue = calculate_evalue(max_score, len(seq1), len(seq2), lambda_value)
    print(round((fim - inicio), 2))
    print(max_score)
    print(barrinha.count('-'))
    print(round(evalue,2))
    print("59")
seq1 = sys.argv[1]
seq2 = sys.argv[2]
lambda_value = 9.162242926908048
smith_waterman(seq1, seq2, lambda_value=lambda_value)